### PR TITLE
Client controller broker selector

### PR DESCRIPTION
**Pull request by saurabhd336**
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        @JsonProperty("port")
+        public int _port;
+
+        @JsonProperty("instanceName")
+        public String _instanceName;
+    }
+
+    private class BrokerUpdater extends Thread {
+        private final AsyncHttpClient _client;
+        private static final String ADDRESS_FORMAT = "http://%s:%d/v2/brokers/tables";
```
Comment (mayankshriv) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Have a broker url builder util.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        private final ObjectMapper _objectMapper = new ObjectMapper();
```
Comment (mayankshriv) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Use static final as it is a very expensive object.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                Map<String, List<BrokerInstance>> responses = _objectMapper.readValue(responseBody, _responseTypeRef);
+                List<String> brokers = new ArrayList<>();
+                for (Map.Entry<String, List<BrokerInstance>> b: responses.entrySet()) {
+                    b.getValue().forEach(br -> {
+                        brokers.add(br._host + ":" + br._port);
+                    });
+                }
+                _brokers = brokers;
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (mayankshriv) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Are we swallowing exception here?

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            new TypeReference<Map<String, List<BrokerInstance>>>() { };
+
+
+        public BrokerUpdater(String controllerHost, int controllerPort) {
+            _client = Dsl.asyncHttpClient();
+            _address = String.format(ADDRESS_FORMAT, controllerHost, controllerPort);
+        }
+
+        public void updateBrokers() {
+            BoundRequestBuilder getRequest = _client.prepareGet(_address);
```
Comment (mayankshriv) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Break it down into util methods/classes that generate urls for specific requests, and make the http call. Also what about https?

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                e.printStackTrace();
+            }
+        }
+
+        @Override
+        public void run() {
+            while (true) {
+                try {
+                    // TODO can we make this reactive instead of polling? websocket endpoint in controller?
+                    Thread.sleep(1000);
```
Comment (mayankshriv) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Make the sleep configurable.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Random;
+import java.util.concurrent.Future;
+import org.asynchttpclient.AsyncHttpClient;
+import org.asynchttpclient.BoundRequestBuilder;
+import org.asynchttpclient.Dsl;
+import org.asynchttpclient.Response;
+
+
+// TODO maintain Map of table -> list of brokers, instead of single list of brokers
+public class ControllerBrokerSelector implements BrokerSelector {
```
Comment (mayankshriv) [inline pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Javadoc for the class. Also, if the idea is to get the broker request from the controller, then better to name it something like `PollingBasedBrokerSelector` or similar. And can we model it as a cache that is updated in the background. The update mechanism can also be an interface (with polling, manual trigger, etc as being different implementations).

Comment (saurabhd336) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Ack.
Have changed the design a bit. Added interface for manually / scheduled broker cache updates.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java (hunk 1 above)
Comment (saurabhd336) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Ack

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java (hunk 2 above)
Comment (saurabhd336) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Ack

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java (hunk 4 above)
Comment (saurabhd336) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Ack. Added scheme that is passed to constructor

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java (hunk 3 above)
Comment (saurabhd336) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Ack.
Intellij defaults :)
Added logging wherever needed.

File: pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java (hunk 5 above)
Comment (saurabhd336) [reply pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/ControllerBrokerSelector.java]: Ack. Passed to scheduledUpdater's constructor

**Pull request closed without merging**